source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -84,5 +106,67 @@ std::vector<Address::InstanceConstSharedPtr> HappyEyeballsConnectionProvider::so
   return address_list;
 }
 
+std::vector<Address::InstanceConstSharedPtr>
+HappyEyeballsConnectionProvider::sortAddressesWithConfig(
+    const std::vector<Address::InstanceConstSharedPtr>& in,
+    const envoy::config::cluster::v3::Cluster::HappyEyeballsConfig& happy_eyeballs_config) {
+  ENVOY_LOG_EVENT(debug, "happy_eyeballs_sort_address", "sort address with happy_eyeballs config.");
+  std::vector<Address::InstanceConstSharedPtr> address_list;
+  address_list.reserve(in.size());
+
+  // Iterator which will advance through all addresses matching the first
+  // family.
+  auto first = in.begin();
+  // Iterator which will advance through all addresses not matching the first
+  // family. This initial value is ignored and will be overwritten in the loop
+  // below.
+  auto other = in.begin();
+
+  // First_family_ip_version is default to the first valid ip version
+  // unless overwritten by happy_eyeballs_config.
+  Address::IpVersion first_family_ip_version = in[0].get()->ip()->version();
+
+  int first_address_family_count = happy_eyeballs_config.first_address_family_count();
+  switch (happy_eyeballs_config.first_address_family_version()) {
+  case envoy::config::cluster::v3::Cluster::DEFAULT:
+    break;
+  case envoy::config::cluster::v3::Cluster::V4:
+    first_family_ip_version = Address::IpVersion::v4;
+    break;
+  case envoy::config::cluster::v3::Cluster::V6:
+    first_family_ip_version = Address::IpVersion::v6;
+    break;
+  default:
+    break;
+  }
+
+  while (first != in.end() && first->get()->ip()->version() != first_family_ip_version) {
+    first++;
+  }
+  while (other != in.end() && other->get()->ip()->version() == first_family_ip_version) {
+    other++;
+  }
+
+  // Address::IpVersion first_family_ip_version(Address::IpVersion::v4);
+  while (first != in.end() || other != in.end()) {
+    int count = 0;
+    while (first != in.end() && ++count <= first_address_family_count) {
```

Comment:
what happens when first_address_family_count is zero?

Reply:
if count=0, the return address list will contain a list of other ip version, then the first ip type. It doesn't raise an error but doesn't make sense realistically. I will add an assertion that count >= 1 at config generation step.